lib/ut/coroutine2.c

```diff
@@ -0,0 +1,266 @@
+
+	m0_semaphore_down(&ready);
+}
+
+static int coroutine_fom_tick(struct m0_fom *fom, int *x, int *__unused)
+{
+	int key = 1;
+	int val = 2;
+	int rc;
+
+	m0_co_op_reset(OP(fom));
+
+	M0_CO_START(CO(fom));
+	crud(fom, &fom_tree, key, val);
+	rc = M0_CO_END(CO(fom));
+
+	if (rc == -EAGAIN) /* m0_co_op case */
+		return m0_co_op_tick_ret(OP(fom), fom, m0_fom_phase(fom));
+	else if (M0_IN(rc, (M0_FSO_WAIT, M0_FSO_AGAIN)))  /* long_lock case */
+		return rc;
+
+	m0_semaphore_up(&ready);
+	return -1;
+}
+
+static void crud(struct m0_fom *fom, struct test_tree *tree, int k, int v)
+{
+	M0_CO_REENTER(CO(fom),
+		      struct m0_long_lock_link llink;
+		);
+
+	M0_LOG(M0_DEBUG, "tree=%p tree.delete=%d, tree.insert=%d k=%d",
+	       tree, tree->delete, tree->insert, k);
+
+	/*
+	 * This part pretends to take a long lock. In the reality it takes it
+	 * only once, no concurrency expected. Use the following code as an
+	 * example to implement your own.
+	 */
+	m0_long_lock_link_init(&F(llink), fom, NULL);
+	M0_CO_YIELD_RC(CO(fom),
+		       LR(m0_long_write_lock(&fom_lock,
+					     &F(llink),
+					     m0_fom_phase(fom))));
+
+	m0_long_write_unlock(&fom_lock, &F(llink));
+	m0_long_lock_link_fini(&F(llink));
+
+	/*
+	 * Updates pretended tree object asuming two blocking points to simulate
+	 * paged inside insert() and delete().
+	 */
+	M0_CO_FUN(CO(fom), insert(fom, tree, k, v));
+	M0_CO_FUN(CO(fom), delete(fom, tree, k));
+	M0_UT_ASSERT(tree->delete + tree->insert == 22);
+}
+
+static void insert(struct m0_fom *fom, struct test_tree *tree, int k, int v)
+{
+	M0_CO_REENTER(CO(fom));
+
+	M0_LOG(M0_DEBUG, "tree=%p tree.delete=%d, tree.insert=%d k=%d v=%d",
+	       tree, tree->delete, tree->insert, k, v);
+
+	M0_LOG(M0_DEBUG, "yield");
+	tree->insert++;
+	m0_co_op_active(OP(fom));
+	m0_semaphore_up(&insert1);
+	M0_CO_YIELD(CO(fom));
+
+	M0_LOG(M0_DEBUG, "yield");
+	tree->insert++;
+	m0_co_op_active(OP(fom));
+	m0_semaphore_up(&insert2);
+	M0_CO_YIELD(CO(fom));
+}
+
+static void delete(struct m0_fom *fom, struct test_tree *tree, int k)
+{
+	M0_CO_REENTER(CO(fom));
+
+	M0_LOG(M0_DEBUG, "tree=%p tree.delete=%d, tree.insert=%d",
+	       tree, tree->delete, tree->insert);
+
+	M0_LOG(M0_DEBUG, "yield");
+	tree->delete += 10;
+	m0_co_op_active(OP(fom));
+	m0_semaphore_up(&delete1);
+	M0_CO_YIELD(CO(fom));
+
+	M0_LOG(M0_DEBUG, "yield");
+	tree->delete += 10;
+	m0_co_op_active(OP(fom));
+	m0_semaphore_up(&delete2);
+	M0_CO_YIELD(CO(fom));
+}
+
+static void test_run(void)
+{
+	struct m0_reqh *reqh = m0_ut__reqh_init();
+	(void) reqh;
+
+	M0_SET0(&fom_simple);
+	M0_FOM_SIMPLE_POST(&fom_simple, reqh, &coroutine2_conf,
+			   &coroutine_fom_tick, NULL, NULL, 1);
+	coroutine_fom_run();
+	m0_ut__reqh_fini();
+}
+
+void m0_test_coroutine2(void)
+{
+	int rc;
+
+	m0_semaphore_init(&ready, 0);
+	m0_semaphore_init(&insert1, 0);
+	m0_semaphore_init(&insert2, 0);
+	m0_semaphore_init(&delete1, 0);
+	m0_semaphore_init(&delete2, 0);
+	m0_long_lock_init(&fom_lock);
+
```

Comment:
Please note, the variable "fom_tree" must also be initialized here.
Otherwise, when we pass "-n 2" or any number more than 2 to run-ut, it will panic.
Because the stats are not cleared.
For example, try this:
```
./scripts/m0 run-ut -t libm0-ut -n 3
```

Reply:
Thanks!
Updated.

Reply:
OK